include/boost/math/distributions/kolmogorov_smirnov.hpp

```diff
@@ -0,0 +1,494 @@
+// Equation 2.4.4 of
+//
+// Durbin J (1973). “Distribution Theory for Tests Based on the Sample
+// Distribution Func- tion.” In SIAM CBMS-NSF Regional Conference Series in
+// Applied Mathematics. SIAM, Philadelphia, PA.
+//
+// which is available in book form from Amazon and others. This exact version
+// involves taking powers of large matrices. To do that right you need to
+// compute eigenvalues and eigenvectors, which are beyond the scope of Boost.
+// (Some recent work indicates the exact form can also be computed via FFT, see
+// https://cran.r-project.org/web/packages/KSgeneral/KSgeneral.pdf).
+//
+// Even if the CDF of the exact distribution could be computed using Boost
+// libraries (which would be cumbersome), the PDF would present another
+// difficulty. Therefore I am limiting this implementation to the asymptotic
+// form, even though the exact form has trivial values for certain specific
+// values of x and n. For more on trivial values see
+//
+// Ruben H, Gambino J (1982). “The Exact Distribution of Kolmogorov’s Statistic
+// Dn for n ≤ 10.” Annals of the Institute of Statistical Mathematics.
+// 
+// For a good bibliography and overview of the various algorithms, including
+// both exact and asymptotic forms, see
+// https://www.jstatsoft.org/article/view/v039i11
+//
+// As for this implementation: the distribution is parameterized by n (number
+// of observations) in the spirit of chi-squared's degrees of freedom. It then
+// takes a single argument x. In terms of the Kolmogorov-Smirnov statistical
+// test, x represents the distribution of D_n, where D_n is the maximum
+// difference between the CDFs being compared, that is,
+//
+//   D_n = sup|F_n(x) - G(x)|
+//
+// In the exact distribution, x is confined to the support [0, 1], but in this
+// limiting approximation, we allow x to exceed unity (similar to how a normal
+// approximation always spills over any boundaries).
+//
+// As mentioned previously, the CDF is implemented using the \tau
+// parameterization of the fourth Jacobi Theta function as
+//
+// CDF=θ₄(0|2*x*x*n/pi)
+//
+// The PDF is a hand-coded derivative of that function. Actually, there are two
+// (independent) derivatives, as separate code paths are used for "small x"
+// (2*x*x*n < pi) and "large x", mirroring the separate code paths in the
+// Jacobi Theta implementation to achieve fast convergence. Quantiles are
+// computed using a Newton-Raphson iteration from an initial guess that I
+// arrived at by trial and error.
+//
+// The mean and variance are implemented using simple closed-form expressions.
+// The mode is calculated at run-time by maximizing the PDF. If you have an
+// analytical solution for the mode, feel free to plop it in.
+//
+// The CDF and PDF could almost certainly be re-implemented and sped up using a
+// polynomial or rational approximation, since the only meaningful argument is
+// x * sqrt(n). But that is left as an exercise for the next maintainer.
+//
```

Comment:
I love these header line docs; really useful for more detail than is appropriate for end-user docs.